pkg/embeddedcluster/node_join.go

```diff
@@ -0,0 +1,339 @@
+	defer joinToken.Mut.Unlock()
+
+	// if the joinToken has been generated in the past hour, return it
+	if joinToken.Creation != nil && time.Now().Before(joinToken.Creation.Add(time.Hour)) {
+		return joinToken.Token, nil
+	}
+
+	newToken, err := runAddNodeCommandPod(ctx, client, nodeRole)
+	if err != nil {
+		return "", fmt.Errorf("failed to run add node command pod: %w", err)
+	}
+
+	now := time.Now()
+	joinToken.Token = newToken
+	joinToken.Creation = &now
+
+	return newToken, nil
+}
+
+// run a pod that will generate the add node token
+func runAddNodeCommandPod(ctx context.Context, client kubernetes.Interface, nodeRole string) (string, error) {
+	podName := "k0s-token-generator-"
+	suffix := strings.Replace(nodeRole, "+", "-", -1)
+	podName += suffix
+
+	// cleanup the pod if it already exists
+	err := client.CoreV1().Pods("kube-system").Delete(ctx, podName, metav1.DeleteOptions{})
+	if err != nil {
+		if !kuberneteserrors.IsNotFound(err) {
+			return "", fmt.Errorf("failed to delete pod: %w", err)
+		}
+	}
+
+	hostPathFile := corev1.HostPathFile
+	hostPathDir := corev1.HostPathDirectory
+	_, err = client.CoreV1().Pods("kube-system").Create(ctx, &corev1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      podName,
+			Namespace: "kube-system",
+			Labels: map[string]string{
+				"replicated.app/embedded-cluster": "true",
+			},
+		},
+		Spec: corev1.PodSpec{
+			RestartPolicy: corev1.RestartPolicyOnFailure,
+			HostNetwork:   true,
+			Volumes: []corev1.Volume{
+				{
+					Name: "bin",
+					VolumeSource: corev1.VolumeSource{
+						HostPath: &corev1.HostPathVolumeSource{
+							Path: "/usr/local/bin/k0s",
+							Type: &hostPathFile,
+						},
+					},
+				},
+				{
+					Name: "lib",
+					VolumeSource: corev1.VolumeSource{
+						HostPath: &corev1.HostPathVolumeSource{
+							Path: "/var/lib/k0s",
+							Type: &hostPathDir,
+						},
+					},
+				},
+				{
+					Name: "etc",
+					VolumeSource: corev1.VolumeSource{
+						HostPath: &corev1.HostPathVolumeSource{
+							Path: "/etc/k0s",
+							Type: &hostPathDir,
+						},
+					},
+				},
+				{
+					Name: "run",
+					VolumeSource: corev1.VolumeSource{
+						HostPath: &corev1.HostPathVolumeSource{
+							Path: "/run/k0s",
+							Type: &hostPathDir,
+						},
+					},
+				},
+			},
+			Affinity: &corev1.Affinity{
+				NodeAffinity: &corev1.NodeAffinity{
+					RequiredDuringSchedulingIgnoredDuringExecution: &corev1.NodeSelector{
+						NodeSelectorTerms: []corev1.NodeSelectorTerm{
+							{
+								MatchExpressions: []corev1.NodeSelectorRequirement{
+									{
+										Key:      "node.k0sproject.io/role",
+										Operator: corev1.NodeSelectorOpIn,
+										Values: []string{
+											"control-plane",
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			Containers: []corev1.Container{
+				{
+					Name:    "k0s-token-generator",
+					Image:   "ubuntu:latest", // TODO use the kotsadm image here as we'll know it exists
```

Comment:
is this a "we'll do this in a separate" pr type of TODO?

Reply:
haaaaaah
no that's a "we intended to be using a different method that didn't happen yet" type of TODO
let me see how much effort that'll take